### PR TITLE
Homogeneous tkv left padded optimization - prefill compute and kv-cache block id re-use

### PR DESCRIPTION
**Pull request by JRosenkranz**
This PR greatly reduces the compute wastage during prefill by only padding to the smallest multiple of 64 that fits the prompt. It also re-uses kv-cache blocks for pads in order to reduce the need for dedicated pad blocks.

### CODE REVIEW

**Comment by JRosenkranz** [comment]
bot:test
TEST_FILE=test_scripts.py

**Comment by JRosenkranz** [comment]
bot:test
TEST_FILE=test_scripts.py
bot:test
TEST_FILE=test_decoders.py MODEL_ID=ibm-granite/granite-3.3-8b-instruct BATCH_SIZE=8 SEQUENCE_LENGTH=64 USE_TINY_MODEL=1

**Comment by JRosenkranz** [comment]
bot:test
TEST_FILE=test_decoders.py MODEL_ID=ibm-granite/granite-3.3-8b-instruct BATCH_SIZE=8 SEQUENCE_LENGTH=2048 USE_TINY_MODEL=1

File: aiu_fms_testing_utils/utils/paged.py [hunk 1]
```diff
@@ -106,21 +106,15 @@ def generate(
...
-            "VLLM_DT_MAX_CONTEXT_LEN",
-            str(
-                (((input_ids.size(1) + max_new_tokens - 1) // BLOCK_SIZE) + 1)
-                * BLOCK_SIZE
-            ),
-        )
-    )
+
+    _MAX_BATCH = int(os.environ["VLLM_DT_MAX_BATCH_SIZE"])
+    _MAX_CONTEXT_LENGTH = int(os.environ["VLLM_DT_MAX_CONTEXT_LEN"])
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/paged.py:115]: add a comment that these are guaranteed to be set somewhere else

File: aiu_fms_testing_utils/utils/paged.py [hunk 2]
```diff
@@ -223,45 +231,75 @@ def generate(
...
         if i > 0:
             kwargs["mask"] = None
             kwargs["position_ids"] = kwargs["position_ids"][:, -1:] + 1
-            pos_i = result.size(1) - 1
-            if pos_i % BLOCK_SIZE == 0:
-                for block_table_i in block_table:
-                    block_number = block_numbers.pop(0)
-                    block_table_i.append(block_number)
-            block_offset = pos_i % BLOCK_SIZE
+            current_tkv_mask = current_tkv_mask + 1
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/paged.py:241]: why do this here and not after lines 245-246 to avoid adding 1 and removing 1?

File: aiu_fms_testing_utils/utils/paged.py [hunk 3]
```diff
@@ -336,7 +377,14 @@ def generate(
             torch._dynamo.mark_static(kwargs["slot_mapping"], 1)  # always 1
             torch._dynamo.mark_static(kwargs["position_ids"], 1)  # always 1
 
-            output = model(input_ids, **kwargs)
+            logits, past_key_value_states = model(input_ids, **kwargs)
+
+            # handle the only_last_token here as it is already being handled above in prefill
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/paged.py:382]: comment is unclear, it looks like something is being done twice

**Review by ani300 — APPROVED** [review]
a bunch of the code should be made clearer, but given the core paged attention tests are passing, it LGTM

File: aiu_fms_testing_utils/utils/paged.py [hunk 4]
```diff
@@ -194,27 +188,41 @@ def generate(
...
+        for pos_i in range(seq_tkv):
+            # we may have already popped a block, so index to the proper block
+            block_number = block_table_i[pos_i // BLOCK_SIZE]
+
             block_offset = pos_i % BLOCK_SIZE
             slot = block_number * BLOCK_SIZE + block_offset
             slot_mapping_i.append(slot)
         slot_mapping.append(slot_mapping_i)
         block_table.append(block_table_i)
-    kwargs["slot_mapping"] = torch.tensor(slot_mapping, dtype=torch.int64)
```
[the remark this answers is not in the document]
Comment (JRosenkranz) [reply aiu_fms_testing_utils/utils/paged.py:214]: the prefill isn't using kwargs, we have a loop there that creates the slot_mapping separately

File: scripts/inference.py [hunk 5]
```diff
@@ -690,6 +690,18 @@ def truncate_prompts_to_max_length(prompts, max_len, max_allowed_length):
...
+if "paged" in attn_name:
+    import bisect
+
+    __largest_context = ids.shape[1] + args.max_new_tokens
+    __supported_context_lengths = [64, 128, 256, 512, 1024, 2048, 4096, 8192]
+    os.environ["VLLM_DT_MAX_CONTEXT_LEN"] = str(
+        __supported_context_lengths[
+            bisect.bisect_left(__supported_context_lengths, __largest_context)
+        ]
+    )
```
[the remark this answers is not in the document]
Comment (JRosenkranz) [reply scripts/inference.py]: fixed in latest commit

File: tests/models/test_decoders.py [hunk 6]
```diff
@@ -162,8 +162,14 @@
...
 if compile_dynamic_sendnn:
+    import bisect
+
+    __largest_context = max(common_seq_lengths) + max(common_max_new_tokens)
+    __supported_context_lengths = [256, 512, 1024, 2048, 4096, 8192]
     os.environ["VLLM_DT_MAX_CONTEXT_LEN"] = str(
-        (((max(common_seq_lengths) + max(common_max_new_tokens)) // 64) + 1) * 64
+        __supported_context_lengths[
+            bisect.bisect_left(__supported_context_lengths, __largest_context)
+        ]
```
[the remark this answers is not in the document]
Comment (JRosenkranz) [reply tests/models/test_decoders.py:174]: fixed in latest commit

File: aiu_fms_testing_utils/utils/paged.py (hunk 3 above)
Comment (JRosenkranz) [reply aiu_fms_testing_utils/utils/paged.py]: added clarification in comments

File: aiu_fms_testing_utils/utils/paged.py (hunk 2 above)
Comment (JRosenkranz) [reply aiu_fms_testing_utils/utils/paged.py]: fixed in latest commit

File: aiu_fms_testing_utils/utils/paged.py (hunk 1 above)
Comment (JRosenkranz) [reply aiu_fms_testing_utils/utils/paged.py:117]: fixed in latest commit

**Comment by JRosenkranz** [comment]
bot:test
TEST_FILE=test_scripts.py

**Comment by JRosenkranz** [comment]
bot:test
TEST_FILE=test_decoders.py MODEL_ID=ibm-granite/granite-3.3-8b-instruct BATCH_SIZE=8 SEQUENCE_LENGTH=2048 USE_TINY_MODEL=1

**Comment by JRosenkranz** [comment]
bot:test
TEST_FILE=test_decoders.py MODEL_ID=ibm-granite/granite-3.3-8b-instruct BATCH_SIZE=8 SEQUENCE_LENGTH=2048 USE_TINY_MODEL=1
